BehaviorTreeEditor.cs

```diff
@@ -9,276 +9,391 @@
 
-	public static TreeNode CreateNode( Type nodeType )
-	{
-		DebugUtils.Assert( _instance, "Cannot create a node without a current editor instance!" );
-		DebugUtils.Assert( _instance._behaviorTree, "Can't create a node without an asset to add it to!" );
+			// Use ID of new node, rather than reusing old node
+			nodeData.id = newNode.id;
 
-		TreeNode newNode = (TreeNode)ScriptableObject.CreateInstance( nodeType );
-		AssetDatabase.AddObjectToAsset( newNode, _instance._behaviorTree );
-		EditorUtility.SetDirty( _instance._behaviorTree );
-		return newNode;
-	}
+			// Give node data a reference to the new node
+			nodeData.node = newNode;
 
-	public static TreeNode nullNode
-	{
-		get
-		{
-			return CreateNode( typeof( NullNode ) );
-		}
-	}
+			// Put new node back in the dict.
+			_behaviorTree._editorData.Add( nodeData );
+			BuildEditorData();
 
-	public static BehaviorTree CloneTree( BehaviorTree behaviorTree )
-	{
-		BehaviorTree treeClone = Instantiate<BehaviorTree>( behaviorTree );
-		treeClone.root = CloneNode( behaviorTree.root );
-		return treeClone;
-	}
+			return newNode;
+		}
 
-	public static TreeNode CloneNode( TreeNode node )
-	{
-		TreeNode nodeClone = Instantiate<TreeNode>( node );
+		public TreeNode nullNode
+		{
+			get
+			{
+				return CreateNodeWithoutParent( typeof( NullNode ) );
+			}
+		}
 
-		if ( node is Decorator )
+		public static BehaviorTree CloneTree( BehaviorTree behaviorTree )
 		{
-			( (Decorator)nodeClone )._child = CloneNode( ( (Decorator)node )._child );
+			BehaviorTree treeClone = Instantiate<BehaviorTree>( behaviorTree );
+			treeClone.root = CloneNode( behaviorTree.root );
+			return treeClone;
 		}
-		else if ( node is Compositor )
+
+		public static TreeNode CloneNode( TreeNode node )
 		{
-			List<TreeNode> cloneChildren = new List<TreeNode>();
-			foreach ( TreeNode child in ( (Compositor)node )._children )
+			TreeNode nodeClone = Instantiate<TreeNode>( node );
+
+			if ( node is Decorator )
+			{
+				( (Decorator)nodeClone )._child = CloneNode( ( (Decorator)node )._child );
+			}
+			else if ( node is Compositor )
 			{
-				cloneChildren.Add( (TreeNode)CloneNode( child ) );
+				List<TreeNode> cloneChildren = new List<TreeNode>();
+				foreach ( TreeNode child in ( (Compositor)node )._children )
+				{
+					cloneChildren.Add( (TreeNode)CloneNode( child ) );
+				}
+
+				( (Compositor)nodeClone )._children = cloneChildren;
 			}
 
-			( (Compositor)nodeClone )._children = cloneChildren;
+			return nodeClone;
 		}
-
-		return nodeClone;
 	}
 }
-
-class EditorNode
+#endif
+namespace BehaviorTree
 {
-	public bool foldout = true;
-}
+	[Serializable]
+	public class EditorData
+	{
+		// NOTE: This should not be called directly.
+		//public EditorData() { }
+
+		public EditorData( TreeNode node, TreeNode parent )
+		{
+			this.node = node;
+			this.parent = parent;
+			this.id = node.id;
+
+			if ( parent is Compositor )
+			{
+				parentIndex = ( (Compositor)parent )._children.Count;
```

Comment:
how is this correct? Is it just by coincidence that the editordata is created just after the node is added to the child list?